src/Microsoft.IdentityModel.JsonWebTokens/JwtTokenUtilities.cs

```diff
@@ -278,6 +284,87 @@ internal static object GetClaimValueUsingValueType(Claim claim)
             return claim.Value;
         }
 
+        /// <summary>
+        /// Gets ValueType of the claim from it's Value.
+        /// </summary>
+        /// <param name="value">Represents value of a claim.</param>
+        /// <returns>String representing claim's ValueType.</returns>
+        internal static string GetClaimValueTypeFromValue(object value)
+        {
+            if (value.GetType().Name == typeof(String).Name && value.ToString() == string.Empty)
+                return JsonClaimValueTypes.JsonNull;
+
+            string claimValueType = TokenUtilities.GetClaimValueTypeFromValue(value);
+
+            if (!String.IsNullOrEmpty(claimValueType))
+                return claimValueType;
+
+            if (value.GetType().Name == typeof(JObject).Name)
+            {
+                return JsonClaimValueTypes.Json;
+            }
+            if (value.GetType().Name == typeof(JArray).Name)
+            {
+                return JsonClaimValueTypes.JsonArray;
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Creats claims from dictionary.
+        /// </summary>
+        /// <param name="claimsCollection">Represents claims stored in dictionary.</param>
+        /// <returns>claims</returns>
+        internal static IEnumerable<Claim> CreateClaimsFromDictionary(IDictionary<string, object> claimsCollection)
+        {
+            List<Claim> claims = new List<Claim>();
+            object value;
+            foreach (string claimtype in claimsCollection.Keys)
+            {
+                claimsCollection.TryGetValue(claimtype, out value);
+                if (value != null)
+                {
+                    string valueType = GetClaimValueTypeFromValue(value);
+
+                    if (value.GetType().Name == typeof(List<>).Name)
+                    {
+                        foreach (var item in (IList)value)
```

Comment:
using `is` with pattern matching directly above will remove the need for a cast here

Reply:
When using` is`, not sure why but JObject and JSONArray also resolve to match IList, which is not expected in this case, as multiple Claims should not be created from a JObject/JSONArray.

Reply:
The why would be that JObject and JSONArray implement IList. That makes sense. Thank you for looking into that.

Reply:
Yes. Indeed it is. Thank you for clearing that up.